### ISSUE TITLE
`textDocument/diagnostic` and `textDocument/semanticTokens/full` not requested after undo (`u`) operation

### ISSUE THREAD

**Issue by soifou** [EmmyLuaLs/emmylua-analyzer-rust#1121]
After an undo operation, the server receives `textDocument/didChange` correctly but does not trigger a refresh of diagnostics or semantic tokens. Only `textDocument/documentColor` and `textDocument/foldingRange` are requested in response. This does not happen with normal edits (insert/delete), where all four requests are correctly triggered after `didChange`.

To reproduce, open a Lua file with 2 lines, e.g.:

```lua
local foo = 1
print(foo)
```

Delete a character with `x`, diagnostics and semantic tokens refresh correctly
Undo with `u`, diagnostics and semantic tokens are not refreshed

Here are the relevant log coming from `vim.lsp` for these two operations:

After `x` (working)
```log
[DEBUG] "rpc.send"	{ jsonrpc = "2.0", method = "textDocument/didChange", params = { contentChanges = { { text = "local oo = {}\nprint(foo)\n" } }, textDocument = { uri = "file://...", version = 26 } } }
[DEBUG]	"LSP[emmylua_ls]"	"client.request"	1	"textDocument/diagnostic"	{ identifier = "EmmyLua", textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG] "rpc.send"	{ id = 78, jsonrpc = "2.0", method = "textDocument/diagnostic", params = { identifier = "EmmyLua", textDocument = { uri = "file://..." } } }
[DEBUG] "LSP[emmylua_ls]"	"client.request"	1	"textDocument/semanticTokens/full"	{ textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG] "rpc.send"	{ id = 79, jsonrpc = "2.0", method = "textDocument/semanticTokens/full", params = { textDocument = { uri = "file://..." } } }
[DEBUG] "LSP[emmylua_ls]"	"client.request"	1	"textDocument/documentColor"	{ textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG]	"rpc.send"	{ id = 80, jsonrpc = "2.0", method = "textDocument/documentColor", params = { textDocument = { uri = "file://..." } } }
[DEBUG]	"rpc.receive"	{ id = 78, jsonrpc = "2.0", result = { items = { { code = "unused", message = "oo is never used, if this is intentional, prefix it with an underscore: _oo", range = { ["end"] = { character = 8, line = 0 }, start = { character = 6, line = 0 } }, severity = 4, source = "EmmyLua", tags = { 1 } }, { code = "undefined-global", message = "undefined global variable: foo", range = { ["end"] = { character = 9, line = 1 }, start = { character = 6, line = 1 } }, severity = 1, source = "EmmyLua" } }, kind = "full" } }
[DEBUG]	"rpc.receive"	{ id = 79, jsonrpc = "2.0", result = { data = { 0, 0, 5, 15, 0, 0, 6, 2, 8, 0, 0, 3, 1, 21, 0, 1, 0, 5, 12, 512, 0, 5, 1, 23, 0, 0, 1, 3, 8, 0, 0, 3, 1, 23, 0 } } }
[DEBUG]	"rpc.receive"	{ id = 80, jsonrpc = "2.0", result = {} }
[DEBUG] "LSP[emmylua_ls]"	"client.request"	1	"textDocument/foldingRange"	{ textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG] "rpc.send"	{ id = 81, jsonrpc = "2.0", method = "textDocument/foldingRange", params = { textDocument = { uri = "file://..." } } }
[DEBUG]	"rpc.receive"	{ id = 81, jsonrpc = "2.0", result = {} }
```

After `u` (broken)
```
[DEBUG]	"rpc.send"	{ jsonrpc = "2.0", method = "textDocument/didChange", params = { contentChanges = { { text = "local foo = {}\nprint(foo)\n" } }, textDocument = { uri = "file://...", version = 27 } } }
[DEBUG]	"LSP[emmylua_ls]"	"client.request"	1	"textDocument/documentColor"	{ textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG]	"rpc.send"	{ id = 82, jsonrpc = "2.0", method = "textDocument/documentColor", params = { textDocument = { uri = "file://..." } } }
[DEBUG]	"rpc.receive"	{ id = 82, jsonrpc = "2.0", result = {} }
[DEBUG] "LSP[emmylua_ls]"	"client.request"	1	"textDocument/foldingRange"	{ textDocument = { uri = "file://..." } }	<function 1>	1
[DEBUG] "rpc.send"	{ id = 83, jsonrpc = "2.0", method = "textDocument/foldingRange", params = { textDocument = { uri = "file://... " } } }
[DEBUG]	"rpc.receive"	{ id = 83, jsonrpc = "2.0", result = {} }
```

My current workaround is to manually requesting diagnostics and semantic tokens after undo/redo in the LSP `on_attach`:

```lua
on_attach = function(_, bufnr)
  local function refresh()
    vim.lsp.buf_request(bufnr, 'textDocument/diagnostic', {
      identifier = 'EmmyLua',
      textDocument = { uri = vim.uri_from_bufnr(bufnr) },
    })
    vim.lsp.semantic_tokens.force_refresh(bufnr)
  end

  local opts = { buffer = bufnr }
  vim.keymap.set('n', 'u', function()
    vim.cmd('undo')
    vim.schedule(refresh)
  end, opts)

  vim.keymap.set('n', '<C-r>', function()
    vim.cmd('redo')
    vim.schedule(refresh)
  end, opts)
end,
```

Notes that `lua_ls` does not exhibit this behavior, undo correctly refreshes diagnostics and semantic tokens.

The server does receive `textDocument/didChange` on undo (confirmed via LSP log), so this is not a client-side issue. It appears to treat undo-originated didChange notifications differently from insert/delete-originated ones, possibly a bug in its change detection or document invalidation logic.

I am running `emmylua-analyzer-rust` version `0.23.1` (not tried the latest because I get hang issues)

**Comment by CppCXY** [comment]
For language servers that support clients with textDocument/diagnostic, the server will use textDocument/diagnostic specifically to handle document diagnostic updates, and will no longer automatically attach a pushDiagnostic based on document changes. lua_ls has not yet updated to the LSP 3.18 specification, so it can still work for now.

**Comment by CppCXY** [comment]
If the client does not send textDocument/diagnostic and semanticToken after an undo, that should be a bug. I tested this in VSCode and found that diagnostics do update promptly after an undo.

**Comment by soifou** [comment]
Thanks for the answer. Actually, this is indeed more of an issue with Neovim. I found a related issue [here](https://github.com/neovim/neovim/issues/32247). Other "pull-diagnostic" servers such as `ty` or `jsonls` have the same problem with stale diagnostics, so I'm closing this issue.
